core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1400,40 +1668,48 @@ object IsrChangeNotificationListener {
  * Starts the preferred replica leader election for the list of partitions specified under
  * /admin/preferred_replica_election -
  */
-class PreferredReplicaElectionListener(protected val controller: KafkaController) extends ControllerZkDataListener {
-  private val controllerContext = controller.controllerContext
+class PreferredReplicaElectionListener(controller: KafkaController) extends IZkDataListener with Logging {
+  override def handleDataChange(dataPath: String, data: Any): Unit = {
+    val partitions = PreferredReplicaLeaderElectionCommand.parsePreferredReplicaElectionData(data.toString)
+    controller.addToControllerEventQueue(controller.PreferredReplicaLeaderElection(partitions))
+  }
 
-  protected def logName = "PreferredReplicaElectionListener"
+  override def handleDataDeleted(dataPath: String): Unit = {}
+}
+
+class ControllerChangeListener(controller: KafkaController) extends IZkDataListener {
```

Comment:
We can do this in a separate PR, but we should consider moving all these listeners to a separate file. This file has almost 2k lines of code and it doesn't seem necessary.